src/ecs.ts

```diff
@@ -0,0 +1,59 @@
+import { ComponentIter } from './ecs';
+import { transpose } from 'ramda';
+
+
+export class Entity {
+  components: Record<string, number> = {};
+}
+
+export class Component {}
+export type ComponentIter<T> = T[];
+
+export interface System <T extends Component[]> {
+  components: Component[];
+  onUpdate(componentIter: ComponentIter<T>);
+}
+
+export class ECS {
+  public entities: Entity[] = [];
+  public components: Record<string, Component[]> = {};
+  public systems: System<any>[] = [];
+
+  public createEntity(): Entity {
+    const entity = new Entity();
+    this.entities.push(entity);
+    return entity;
+  }
+
+  public registerComponent(component: Component): ECS {
+    this.components[component.constructor.name] = [];
+    return this;
+  }
+
+  public registerSystem <T extends Component[]> (system: System<T>): ECS {
+    this.systems.push(system);
+    return this;
+  }
+
+  public addComponent(entity: Entity, component: Component): Entity {
+    const name = component.constructor.name;
+
+    if (!(name in this.components)) {
+      this.registerComponent(component);
+    }
+
+    const index = this.components[name].push(component) - 1;
+    entity.components[component.constructor.name] = index;
+    return entity;
+  }
+
+  public run() {
+    for (const system of this.systems) {
+      const componentIter = transpose(system.components.map(componentClass => {
```

Comment:
This makes a dangerous assumption that given a ecs that has a `System<[A, B]>` that all entities that have component `A` will also have component `B`.

In other words, I was able to break the test by adding a second entity with only a position.

See the branch `systems-cause-entity-requirements`, specifically https://github.com/ryanhinkel/pixistorm/blob/systems-cause-entity-requirements/src/ecs.test.ts#L77-L78